### PR TITLE
Add ability to extract CSS from SFC

### PR DESCRIPTION
**Pull request by dwightjack**
This PR should address https://github.com/apeschar/esbuild-vue/issues/5.

I am not really familiar with ESBuild and Vue compiler, so I used the svelte plugin implementation as reference (https://github.com/EMH333/esbuild-svelte/blob/main/index.ts#L68).

Basically by setting the new `extractCSS: true` option the plugin will create a pseudo-file for each compiled style block and pass it to ESBuild CSS loader. In this way, ESBuild will generate a CSS file with the same filename of the JS output. 

I also added a `production` option that will force Vue compiler in production mode (basically removes the SFC source from the bundle).

Finally I added an example usage and prettier to give the codebase some consistency 😊

### CODE REVIEW

File: src/worker.js [hunk 1]
```diff
@@ -25,30 +25,50 @@ editModule("fs", (fs) => {
...
+    if (extractCss) {
+      styles = result.styles.map(({ code, scoped, media, module }) => ({
+        code,
+        scoped,
+        media,
+        module,
+      }));
+      result.styles = result.styles.map((style) =>
+        Object.assign(style, { code: "" })
+      );
```
Comment (apeschar) [inline src/worker.js:60]: What's the reason to not just do `result.styles = []` here?

File: src/worker.js [hunk 2]
```diff
@@ -25,30 +25,50 @@ editModule("fs", (fs) => {
...
-      return { result: { errors: resultErrors }, usedFiles };
+      return { errors: resultErrors, usedFiles };
+    }
+    if (extractCss) {
+      styles = result.styles.map(({ code, scoped, media, module }) => ({
+        code,
+        scoped,
+        media,
+        module,
+      }));
```
Comment (apeschar) [inline src/worker.js:57]: I think we need only `code`? In that case can both object patterns be changed to `{code}`?

**Review by apeschar — CHANGES_REQUESTED** [review]
Very nice!  Thanks a lot!

I have some small comments.  Would you mind looking into those before I merge this?

File: src/worker.js (hunk 2 above)
Comment (dwightjack) [reply src/worker.js:57]: At the moment this PR support only global and scoped CSS but I think that the `media` and `module` property might be used when setting media query for the style block and CSS Modules. I have not tested those two scenarios though.

File: src/worker.js (hunk 1 above)
Comment (dwightjack) [reply src/worker.js]: If we remove the style items, the rendered Vue components will lose their scopeId information. I found that by only clearing the `code` property (which contains the CSS code) scopeId is preserved.

I've optimized a bit the code and added a short comment about this in https://github.com/apeschar/esbuild-vue/pull/6/commits/b793f4a609a6681f2982396dddb8c873549131df

File: src/index.js [hunk 3]
```diff
@@ -20,20 +25,56 @@ module.exports = function ({ workers = true, onReadFile } = {}) {
...
-        const { result, usedFiles } = await runTask({
+        let { code, styles, errors, usedFiles } = await runTask({
           filename,
           source,
+          extractCss,
+          production,
         });
+
+        if (extractCss && Array.isArray(styles) && styles.length) {
+          let cssPath = filename
```
[the remark this answers is not in the document]
Comment (dwightjack) [reply src/index.js]: Fixed in https://github.com/apeschar/esbuild-vue/pull/6/commits/5c32f58baa745d1cef44b66d45e520a948c1e087

File: src/index.js [hunk 4]
```diff
@@ -20,20 +25,56 @@ module.exports = function ({ workers = true, onReadFile } = {}) {
...
+            .replace(".vue", ".esbuild-vue-css")
+            .replace(/\\/g, "/");
+          cssCode.set(
+            cssPath,
+            styles.reduce((str, { code }) => str + code, "")
+          );
+          code = code + `\nimport "${cssPath}";`;
```
[the remark this answers is not in the document]
Comment (dwightjack) [reply src/index.js]: Thanks! fixed in https://github.com/apeschar/esbuild-vue/pull/6/commits/5c32f58baa745d1cef44b66d45e520a948c1e087

**Comment by dwightjack** [comment]
@apeschar Thank you for your review. As I wrote in another comment, at the moment this supports only global and scoped CSS extraction, I am not sure if it's working with CSS Modules or media queries (I believe it has something to do with this loader: https://github.com/ktsn/vue-media-loader).

File: src/worker.js (hunk 2 above)
Comment (apeschar) [reply src/worker.js:57]: I suppose the media query on the style block might be translated to a `@media ... { ... }` block wrapped around the exported CSS.

I'm not familiar with CSS modules and how they're used in inline style blocks in Vue.

**Comment by apeschar** [comment]
I've merged this and released version 0.4.0 including this change. It won't work with media queries. But that's something for another PR.

Thanks for your contribution.

**Comment by dwightjack** [comment]
Great ☺️. I agree, this PR should cover the basic. When I have a little more time I will try as well to explore media queries and CSS modules.

**Pull request merged**